feat(ViewportActionMenu): window level per viewport / new patient info / colorbars/ 3D presets and 3D volume rendering

IbrahimCSAE opened the pull request:
👏👏👏 Big shoutout to @jbocce for kicking off the work on this fantastic feature! 💯 👏👏

---



## ViewportActionMenu, Patient Info Header, 3D Presets and more

This PR will address several feature requests with and design changes

- [x] #3742
- [x] #3743
- [x] #3764
- [x] #3825 
- [x] #3814

Major changes are:

- New action menu per viewport that allows user to control different things including: Window Level, Colormaps, Colorbars, 3D Presets, 3D Rendering Properties
- Addition of new patient info header in the viewer that shows the information of the currently displayed patient
- Introduction of Colorbars to OHIF
- Introduction of 3D Rendering and features allowing to set cinematic rendering, shift, shade, ambiance, diffuse and specular.

## Viewport Action Menu

A fully customizable action menu, where you can define your own window level presets per modality, colormaps, 3D presets.


![image](https://github.com/OHIF/Viewers/assets/93064150/48967615-d71d-4773-b2ec-60407bf24abb)

![image](https://github.com/OHIF/Viewers/assets/93064150/ff9d0ff3-9fca-444b-b60b-fe06097613dd)

Customizing the existing presets can be done using the `customizationService`:

An example below:

```js
function getCustomizationModule() {
  return [
    {
      name: 'default',
      value: [
        {
          id: 'cornerstone.windowLevelPresets',
          presets: defaultWindowLevelPresets,
        },
        {
          id: 'cornerstone.colorbar',
          width: '16px',
          colorbarTickPosition: 'left',
          colormaps,
          colorbarContainerPosition: 'right',
          colorbarInitialColormap: DefaultColormap,
        },
        {
          id: 'cornerstone.3dVolumeRendering',
          volumeRenderingPresets: VIEWPORT_PRESETS,
          volumeRenderingQualityRange: {
            min: 1,
            max: 4,
            step: 1,
          },
        },
      ],
    },
  ];
}
```

For window level, the array of presets follows this structure:
```js
const defaultWindowLevelPresets = {
  CT: [
    { description: 'Soft tissue', window: '400', level: '40' },
    { description: 'Lung', window: '1500', level: '-600' },
    { description: 'Liver', window: '150', level: '90' },
    { description: 'Bone', window: '2500', level: '480' },
    { description: 'Brain', window: '80', level: '40' },
  ],

  PT: [
    { description: 'Default', window: '5', level: '2.5' },
    { description: 'SUV', window: '0', level: '3' },
    { description: 'SUV', window: '0', level: '5' },
    { description: 'SUV', window: '0', level: '7' },
    { description: 'SUV', window: '0', level: '8' },
    { description: 'SUV', window: '0', level: '10' },
    { description: 'SUV', window: '0', level: '15' },
  ],
};
```
You can add the modalities you require, and the window/level values for each preset.


## Colormaps

You can now change the colormap for the viewport using the action menu, not only that, you are also able to change colormaps on fusion viewports, and pick which modality you'd like to change the colormap for.

#### Regular viewport:

![Peek 2024-03-28 06-37](https://github.com/OHIF/Viewers/assets/93064150/c8c4d53e-8726-4dda-b6d0-458b9ff7e909)

#### Fusion viewport:

![Peek 2024-03-28 06-39](https://github.com/OHIF/Viewers/assets/93064150/d2f1074f-ab1a-4b6b-8071-8d0cdcabf65d)


#### Colormaps Previews 
You can also preview colormaps before selecting them, by toggling the preview colormap toggle, which gives you a chance to see if the colormap works for you before selecting it:

![Peek 2024-03-28 06-40](https://github.com/OHIF/Viewers/assets/93064150/ca98e8ff-443f-41fb-8d7d-8e188ddf31fd)

## Colorbars

Colorbars are being introduced to OHIF. They are interactive and can be used to manipulate the VOI on stack and volume viewports. They also support fusion viewports with PT/CT displaysets, and turn into dual mode, each colorbar controlling it's respective displayset.

#### Regular viewport:

![Peek 2024-03-28 06-43](https://github.com/OHIF/Viewers/assets/93064150/96d105d1-5935-4080-91e3-770e4f15d588)

#### Fusion viewport:

![Peek 2024-03-28 06-45](https://github.com/OHIF/Viewers/assets/93064150/fdf6f884-01c5-49db-82c9-1c0d0b4ed347)

#### Responsive to colormap changes
 
Colorbars also respond to colormap changes, as seen here:

![Peek 2024-03-28 06-46](https://github.com/OHIF/Viewers/assets/93064150/142af40b-1b7a-4498-b90b-41db14bb2c2f)

The colorbar state is maintained during the lifecycle of the viewport.

## 3D Rendering

The viewport action menu also allows you to manipulate the currently displayed 3D volume, offering a varity of options such as presets, quality, and more. I will showcase each feature below:

#### Rendering presets

![Peek 2024-03-28 06-53](https://github.com/OHIF/Viewers/assets/93064150/878319f5-bd78-4aa0-88d6-2a3cd35afa27)

There's several presets to choose from, depending on your use case.

#### Rendering Quality

![ezgif-7-36258c8160](https://github.com/OHIF/Viewers/assets/93064150/787931a4-6e6c-43c9-85e3-9e35e5a583a9)

#### Shift

![shift-ezgif com-optimize](https://github.com/OHIF/Viewers/assets/93064150/b71404e5-03c6-4520-93bf-029586a53bb9)

#### Shade

![shade](https://github.com/OHIF/Viewers/assets/93064150/c37aaa2f-66b7-4ced-82ce-174f1bfb0282)

#### Ambient

![ambient](https://github.com/OHIF/Viewers/assets/93064150/7a46bb82-35c5-4488-88fd-85e7bd763858)

#### Diffuse

![diffuse](https://github.com/OHIF/Viewers/assets/93064150/d144d8bf-7213-4749-8cf0-893bcacfc9cb)


#### Specular

![specular](https://github.com/OHIF/Viewers/assets/93064150/537d65f9-6edc-472d-ab3d-f5f6d53fe84b)


## New Patient Info Header

The new patient info header displays information about the patient for the currently displayed study.

![image](https://github.com/OHIF/Viewers/assets/93064150/1cfb4889-06a5-4d0b-ac5d-930c93cc1276)

It also includes safe guards where there are multiple patients loading into the viewer, if the following case is detected, the header will only render a text that indicates there's multiple patients, to avoid assuming it's a single patient.

![image](https://github.com/OHIF/Viewers/assets/93064150/b153ff21-019b-4b03-b1c6-30b38856e3f5)


The patient info header contains 3 configuration options that you can pick from and set in your `appConfig` located in `default.js`

```js
export enum PatientInfoVisibility {
  VISIBLE = 'visible',
  VISIBLE_COLLAPSED = 'visibleCollapsed',
  DISABLED = 'disabled',
}
```
`visible`: the patient info header will always be visible and it's in expanded state.
`visibleCollapsed`: the patient info header will be visible, but it's initial state is collapsed, a user would have to click on it to expand it.
`disabled`: the patient info header will not be shown.

sedghi reviewed (commented):
I don't think we need the dropdown for the window level anymore, just use a simple radioButton like zoom or pan

<img width="430" alt="CleanShot 2024-03-27 at 11 20 42@2x" src="https://github.com/OHIF/Viewe



----

The issue i sent over slack for color bar sticking around with old colors



https://github.com/OHIF/Viewers/assets/7490180/1c339fc0-98e2-43b8-9c6d-02033e9b0302

---

I don't think the icon is correct color in the inverted viewports

<img width="380" alt="CleanShot 2024-03-27 at 11 24 23@2x" src="https://github.com/OHIF/Viewers/assets/7490180/21cf2fa4-bfcd-453c-ae1e-fcebf15325cc">

it probably should be the same color as the text, see text overlay on how we detect background is inverted and accordingly use a separate color for texts

----


when inverted color the colorbar tickers should be black or something they are white seems like , it is not readable

<img width="103" alt="CleanShot 2024-03-27 at 11 25 39@2x" src="https://github.com/OHIF/Viewers/assets/7490180/2f2548e8-4da9-47ef-a3a4-2bfce95101bc">


----

we should probably have another toggle that is invert background so that when the LUT is changed, the user can toggle the background to black to make it work 

<img width="227" alt="CleanShot 2024-03-27 at 11 26 53@2x" src="https://github.com/OHIF/Viewers/assets/7490180/8cf542ee-e455-4edd-9bc4-62468a4b8e3d">

----

something weird happens when you set PT preset, seems like the winodw level does not work after that. maybe it is related to the weird binary image after the preset selection. 

https://github.com/OHIF/Viewers/assets/7490180/f8980ffd-7716-47e2-ae2a-ee17d143029b

----

Not sure why we don't show the dates for the cor, sag ct 

<img width="1153" alt="CleanShot 2024-03-27 at 11 29 42@2x" src="https://github.com/OHIF/Viewers/assets/7490180/5ab20862-c619-4168-aae8-914c560b9615">


----

we should not show I: for the reconstructed views since they are not `instance` number

![CleanShot 2024-03-27 at 11 30 11@2x](https://github.com/OHIF/Viewers/assets/7490180/97ac0064-6532-4fed-b10d-28c34dfddbca)

----

I should not be able to select overlay texts 

<img width="810" alt="CleanShot 2024-03-27 at 11 30 49@2x" src="https://github.com/OHIF/Viewers/assets/7490180/d41dc4a3-59e0-4912-be51-c21ac8f37255">

----

we don't need PT fusion color map after this PR and also remove the dropdown for window level presets for tmtv as well

<img width="544" alt="CleanShot 2024-03-27 at 11 31 37@2x" src="https://github.com/OHIF/Viewers/assets/7490180/0584f069-36ae-4aa4-b3a1-a0c90a84f515">


----

color bar for tmtv is weird 

<img width="774" alt="CleanShot 2024-03-27 at 11 32 45@2x" src="https://github.com/OHIF/Viewers/assets/7490180/ae2ad580-5182-4f25-a87a-485ef6a3be89">


----

logo seems like is a bit to the right, ask Dan, IMO it should align with the left of the viewport grid 

![CleanShot 2024-03-27 at 11 33 06@2x](https://github.com/OHIF/Viewers/assets/7490180/d76a02d0-81c4-482e-a60c-1a4f11eacb77)

similar to how the right side is aligned 

<img width="198" alt="CleanShot 2024-03-27 at 11 33 51@2x" src="https://github.com/OHIF/Viewers/assets/7490180/a07c8608-5468-40e9-8dd1-b1d187747775">

the right side vertical bar also seems black, maybe is wrong? ask Dan
----

(Not your PR) can we fix this as well

<img width="277" alt="CleanShot 2024-03-27 at 11 34 33@2x" src="https://github.com/OHIF/Viewers/assets/7490180/e4731389-8f25-42c1-b030-35ede4b84564">

---

IbrahimCSAE commented:
Fixed

- Colorbar sticking around for new displaysets
- OHIF logo position
- Viewport overlay no longer selectable, Viewport Markers no longer selectable
- Removed PET colormap tool in TMTV
- Removed WindowLevel presets in all modes
- Fixed colorbar position in TMV mode
- Fixed study date not showing up in MPR

Remaining

- instance indicator (i) should not be visible for reconstructed volumes
- black markers for colorbar on inverted displaysets
- inverted icon for window level menu on inverted displaysets


Comments

- For PT window level, it seems the problem is the preset, it puts them in a level where window tool can't control them anymore, u can verify this if you switch back to default preset (first item in list), the window tool will start working again, tell me what u want to do for this
- 
![image](https://github.com/OHIF/Viewers/assets/93064150/ecfeb9aa-e024-4ee8-a243-65f4858f5a77)

I don't understand what needs to be fixed?

- I noticed something has broken the colormap sync again recently in OHIF, it affected this PR as well

https://viewer-dev.ohif.org/tmtv?StudyInstanceUIDs=1.3.6.1.4.1.14519.5.2.1.7009.2403.871108593056125491804754960339

Change the colormap using the tool, then go into 1x1 and comeback out, the sync is broken...

This affected my MPR as well when you change the colormap in none MPR then go into MPR, it goes back to grayscale, it wasn't like this before
